homeassistant/components/mysensors/gateway.py

```diff
@@ -155,68 +167,84 @@ def internal_callback(msg):
+):
     """Load any persistent devices and platforms and start gateway."""
     discover_tasks = []
     start_tasks = []
-    for gateway in gateways.values():
-        discover_tasks.append(_discover_persistent_devices(hass, hass_config, gateway))
-        start_tasks.append(_gw_start(hass, gateway))
+    discover_tasks.append(_discover_persistent_devices(hass, hass_config, gateway))
+    start_tasks.append(_gw_start(hass, gateway))
     if discover_tasks:
         # Make sure all devices and platforms are loaded before gateway start.
         await asyncio.wait(discover_tasks)
     if start_tasks:
         await asyncio.wait(start_tasks)
 
 
-async def _discover_persistent_devices(hass, hass_config, gateway):
+async def _discover_persistent_devices(
+    hass: HomeAssistantType, hass_config: ConfigEntry, gateway: BaseAsyncGateway
+):
     """Discover platforms for devices loaded via persistence file."""
     tasks = []
     new_devices = defaultdict(list)
     for node_id in gateway.sensors:
         if not validate_node(gateway, node_id):
             continue
-        node = gateway.sensors[node_id]
-        for child in node.children.values():
+        node: Sensor = gateway.sensors[node_id]
+        for child in node.children.values():  # child is of type ChildSensor
             validated = validate_child(gateway, node_id, child)
             for platform, dev_ids in validated.items():
                 new_devices[platform].extend(dev_ids)
+    _LOGGER.debug("discovering persistent devices: %s", new_devices)
     for platform, dev_ids in new_devices.items():
-        tasks.append(discover_mysensors_platform(hass, hass_config, platform, dev_ids))
+        discover_mysensors_platform(hass, hass_config, platform, dev_ids)
     if tasks:
         await asyncio.wait(tasks)
 
 
-async def _gw_start(hass, gateway):
+async def gw_stop(hass, gateway: BaseAsyncGateway):
+    """Stop the gateway."""
+    _LOGGER.info("stopping gateway %s", gateway.entry_id)
+    connect_task = hass.data[DOMAIN].get(
+        MYSENSORS_GATEWAY_START_TASK.format(gateway.entry_id), None
+    )
+    if connect_task is not None and not connect_task.done():
+        connect_task.cancel()
+    await gateway.stop()
+
+
+async def _gw_start(hass: HomeAssistantType, gateway: BaseAsyncGateway):
     """Start the gateway."""
     # Don't use hass.async_create_task to avoid holding up setup indefinitely.
-    connect_task = hass.loop.create_task(gateway.start())
+    hass.data[DOMAIN][
+        MYSENSORS_GATEWAY_START_TASK.format(gateway.entry_id)
+    ] = asyncio.create_task(
+        gateway.start()
+    )  # store the connect task so it can be cancelled in gw_stop
 
-    @callback
-    def gw_stop(event):
-        """Trigger to stop the gateway."""
-        hass.async_create_task(gateway.stop())
-        if not connect_task.done():
-            connect_task.cancel()
+    async def stop_this_gw():
```

Comment:
We need to accept an argument which is a core event.
```suggestion
    async def stop_this_gw(_: Event):
```

Reply:
Ah okay i will add that.
Is there no documentation about this or am i just incapable of using a search engine?